iceberg/azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSInputStream.java

```diff
@@ -0,0 +1,203 @@
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.metrics.Counter;
+import org.apache.iceberg.metrics.MetricsContext;
+import org.apache.iceberg.metrics.MetricsContext.Unit;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.io.ByteStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ADLSInputStream extends SeekableInputStream implements RangeReadable {
+  private static final Logger LOG = LoggerFactory.getLogger(ADLSInputStream.class);
+
+  // max skip size, if skip is larger, then the stream will be reopened at new position
+  private static final int SKIP_SIZE = 1024 * 1024;
+
+  private final StackTraceElement[] createStack;
+  private final DataLakeFileClient fileClient;
+  private Long fileSize;
+  private final AzureProperties azureProperties;
+
+  private InputStream stream;
+  private long pos;
+  private long next;
+  private boolean closed;
+
+  private final Counter readBytes;
+  private final Counter readOperations;
+
+  ADLSInputStream(
+      DataLakeFileClient fileClient,
+      Long fileSize,
+      AzureProperties azureProperties,
+      MetricsContext metrics) {
+    this.fileClient = fileClient;
+    this.fileSize = fileSize;
+    this.azureProperties = azureProperties;
+
+    this.readBytes = metrics.counter(FileIOMetricsContext.READ_BYTES, Unit.BYTES);
+    this.readOperations = metrics.counter(FileIOMetricsContext.READ_OPERATIONS);
+
+    this.createStack = Thread.currentThread().getStackTrace();
+
+    openStream();
+  }
+
+  private void openStream() {
+    DataLakeFileOpenInputStreamResult result =
+        fileClient.openInputStream(getInputOptions(new FileRange(pos)));
+    this.fileSize = result.getProperties().getFileSize();
+    this.stream = result.getInputStream();
+  }
+
+  private DataLakeFileInputStreamOptions getInputOptions(FileRange range) {
+    DataLakeFileInputStreamOptions options = new DataLakeFileInputStreamOptions();
+    azureProperties.adlsReadBlockSize().ifPresent(options::setBlockSize);
+    options.setRange(range);
+    return options;
+  }
+
+  @Override
+  public long getPos() {
+    return next;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "Cannot seek: already closed");
+    Preconditions.checkArgument(newPos >= 0, "Cannot seek: position %s is negative", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
+    this.next = newPos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    Preconditions.checkState(!closed, "Cannot read: already closed");
+    positionStream();
+
+    pos += 1;
+    next += 1;
+    readBytes.increment();
+    readOperations.increment();
+
+    return stream.read();
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
```

Comment:
Why is all this code copied? Is it also available in Iceberg library artifacts to use as a library dependency?

Reply:
The submodules is difficult because I need to change a slight thing in the iceberg azure code in order for it to work. I am not going to include them for now but then create a new PR that fixes the root issues with the Tables URI which will then allow me to add the submodules as well.